File: bonsoir_android/android/src/main/kotlin/fr/skyost/bonsoir/broadcast/BonsoirServiceBroadcast.kt
package fr.skyost.bonsoir.broadcast

import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Handler
import android.os.Looper
import android.util.Log
import fr.skyost.bonsoir.BonsoirPlugin
import fr.skyost.bonsoir.BonsoirService
import fr.skyost.bonsoir.SuccessObject
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink

/**
 * Allows to broadcast a NSD service on local network.
 *
 * @param id The listener identifier.
 * @param printLogs Whether to print debug logs.
 * @param onDispose Triggered when this instance is being disposed.
 * @param nsdManager The NSD manager.
 * @param messenger The Flutter binary messenger.
 * @param service The Bonsoir service to broadcast.
 */
class BonsoirServiceBroadcast(
    private val id: Int,
    private val printLogs: Boolean,
    private val onDispose: Runnable,
    private val nsdManager: NsdManager,
    messenger: BinaryMessenger,
    private val service: BonsoirService
) : NsdManager.RegistrationListener {

    /**
     * The current event channel.
     */
    private val eventChannel: EventChannel =
        EventChannel(messenger, "${BonsoirPlugin.channel}.broadcast.$id")

    /**
     * The current event sink.
     */
    private var eventSink: EventSink? = null

    /**
     * Whether the broadcast is active.
     */
    private var isActive = false

    /**
     * Initializes this instance.
     */
    init {
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, eventSink: EventSink) {
                this@BonsoirServiceBroadcast.eventSink = eventSink
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    /**
     * Starts the service registration.
     */
    fun start() {
        if (!isActive) {
            nsdManager.registerService(service.toNsdService(), NsdManager.PROTOCOL_DNS_SD, this)
        }
    }

    override fun onServiceRegistered(service: NsdServiceInfo) {
        isActive = true
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir service registered : ${this.service}")
        }
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(SuccessObject("broadcastStarted", this.service).toJson())
        }
    }

    override fun onRegistrationFailed(service: NsdServiceInfo, errorCode: Int) {
        if (printLogs) {
            Log.d(
                BonsoirPlugin.tag,
                "[$id] Bonsoir service registration failed : ${this.service}, error code : $errorCode"
            )
        }
        Handler(Looper.getMainLooper()).post {
            eventSink?.error("broadcastError", "Bonsoir service registration failed.", errorCode)
        }
        dispose()
    }

    override fun onServiceUnregistered(service: NsdServiceInfo) {
        val wasActive = isActive
        isActive = false
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir service broadcast stopped : ${this.service}")
        }
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(SuccessObject("broadcastStopped", this.service).toJson())
        }
        dispose(wasActive)
    }

    override fun onUnregistrationFailed(service: NsdServiceInfo, errorCode: Int) {
        if (printLogs) {
            Log.d(
                BonsoirPlugin.tag,
                "[$id] Bonsoir service unregistration failed : ${this.service}, error code : $errorCode"
            )
        }
        Handler(Looper.getMainLooper()).post {
            eventSink?.error("broadcastError", "Bonsoir service unregistration failed.", errorCode)
        }
    }

    /**
     * Disposes the current class instance.
     */
    fun dispose(notify: Boolean = isActive) {
        if (isActive) {
            isActive = false
            nsdManager.unregisterService(this)
        }
        if (notify) {
            onDispose.run()
        }
    }
}

File: bonsoir_darwin/darwin/Classes/Broadcast/BonsoirServiceBroadcast.swift
#if canImport(Flutter)
    import Flutter
#endif
#if canImport(FlutterMacOS)
    import FlutterMacOS
#endif
import Network

/// Allows to broadcast a given service to the local network.
@available(iOS 13.0, macOS 10.15, *)
class BonsoirServiceBroadcast: NSObject, FlutterStreamHandler {
    /// The delegate identifier.
    let id: Int

    /// Whether to print debug logs.
    let printLogs: Bool

    /// Triggered when this instance is being disposed.
    let onDispose: (Bool) -> Void
    
    /// The advertised service.
    let service: BonsoirService
    
    /// The reference to the registering..
    var sdRef: DNSServiceRef?
    
    /// The current event channel.
    var eventChannel: FlutterEventChannel?

    /// The current event sink.
    var eventSink: FlutterEventSink?

    /// Initializes this class.
    public init(id: Int, printLogs: Bool, onDispose: @escaping (Bool) -> Void, messenger: FlutterBinaryMessenger, service: BonsoirService) {
        self.id = id
        self.printLogs = printLogs
        self.onDispose = onDispose
        self.service = service
        super.init()
        eventChannel = FlutterEventChannel(name: "\(SwiftBonsoirPlugin.package).broadcast.\(id)", binaryMessenger: messenger)
        eventChannel?.setStreamHandler(self)
    }
    
    func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    /// Starts the broadcast.
    public func start() {
        var txtRecord: TXTRecordRef = TXTRecordRef()
        var txtMac = [Int8](repeating: 0, count: 255)
        if service.attributes != nil {
            TXTRecordCreate(&txtRecord, 255, &txtMac);
            for (key, value) in service.attributes! {
                TXTRecordSetValue(&txtRecord, key, UInt8(value.count), value)
            }
        }
        
        let error = DNSServiceRegister(&sdRef, 0, 0, service.name, service.type, "local.", service.host, UInt16(service.port), UInt16(strlen(txtMac)), txtMac, { sdRef, flags, errorCode, name, regType, domain, context in
            let broadcast = Unmanaged<BonsoirServiceBroadcast>.fromOpaque(context!).takeUnretainedValue()
            if errorCode == kDNSServiceErr_NoError {
                if broadcast.printLogs {
                    SwiftBonsoirPlugin.log(category: "broadcast", id: broadcast.id, message: "Bonsoir service broadcasted : \(broadcast.service.description)")
                }
                broadcast.eventSink?(SuccessObject(id: "broadcastStarted", service: broadcast.service).toJson())
            }
            else {
                if broadcast.printLogs {
                    SwiftBonsoirPlugin.log(category: "broadcast", id: broadcast.id, message: "Bonsoir service failed to broadcast : \(broadcast.service.description), error code : \(errorCode)")
                }
                broadcast.eventSink?(FlutterError.init(code: "broadcastError", message: "Bonsoir service failed to broadcast.", details: errorCode))
                broadcast.dispose()
            }
        }, Unmanaged.passUnretained(self).toOpaque())
        if error == kDNSServiceErr_NoError {
            if printLogs {
                SwiftBonsoirPlugin.log(category: "broadcast", id: id, message: "Bonsoir service broadcast initialized : \(service.description)")
            }
            DNSServiceProcessResult(sdRef)
        } else {
            if printLogs {
                SwiftBonsoirPlugin.log(category: "broadcast", id: id, message: "Bonsoir service failed to broadcast : \(service.description), error code : \(error)")
            }
            eventSink?(FlutterError.init(code: "broadcastError", message: "Bonsoir service failed to broadcast.", details: error))
            dispose()
        }
    }

    /// Disposes the current class instance.
    public func dispose() {
        DNSServiceRefDeallocate(sdRef)
        if printLogs {
            SwiftBonsoirPlugin.log(category: "broadcast", id: id, message: "Bonsoir service broadcast stopped : \(service.description)")
        }
        eventSink?(SuccessObject(id: "broadcastStopped", service: service).toJson())
        onDispose()
    }
}
